### PR TITLE
SD-1873,SD-1866: Normalize JIT

### PR DESCRIPTION
**Pull request by HenrikHL**
### **User description**
[SD-1873](https://dcsa.atlassian.net/browse/SD-1873): Normalize JIT
[SD-1866](https://dcsa.atlassian.net/browse/SD-1866): This also moves properties from PortCall --> TerminalCall

This PR contains very many changes. For a quick review please check this [SwaggerPage](https://app.swaggerhub.com/apis/dcsaorg/DCSA_JIT/2.0.0-pvts#/) that is aligned with this PR

___

### **PR Type**
enhancement, documentation

___

### **Description**
- Introduced new endpoints for managing Port Calls, Terminal Calls, and Timestamps, enhancing the API's functionality.
- Added detailed descriptions and examples for API responses and parameters to improve clarity and usability.
- Introduced new schemas for Port Call Services, Timestamps, and Vessel Statuses, providing a more comprehensive data model.
- Updated filtering logic and descriptions for query parameters to ensure accurate data retrieval.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>JIT_v2.0.0.yaml</strong><dd><code>Extensive API enhancements and documentation updates for JIT v2.0.0</code></dd></summary>
<hr>

jit/v2/JIT_v2.0.0.yaml

<li>Added new endpoints for managing Port Calls, Terminal Calls, and <br>Timestamps.<br> <li> Enhanced descriptions and examples for API responses and parameters.<br> <li> Introduced new schemas for Port Call Services, Timestamps, and Vessel <br>Statuses.<br> <li> Updated filtering logic and descriptions for query parameters.<br>

</details>

  </td>
  <td><a href="https://github.com/dcsaorg/DCSA-OpenAPI/pull/460/files#diff-9515d70681d4860776aad1fa65c622e5b362021ad1f0c0ae1352067f49108299">+1422/-790</a></td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 🔍 To retrieve JIRA tickets as context to your PR description, please [register](https://qodo-merge-docs.qodo.ai/core-abilities/fetching_ticket_context/#jira-cloud) your organization.

[SD-1873]: https://dcsa.atlassian.net/browse/SD-1873?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[SD-1866]: https://dcsa.atlassian.net/browse/SD-1866?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by emiliodcsa — APPROVED** [review]

**Review by emiliodcsa — APPROVED** [review]

**Review by emiliodcsa — APPROVED** [review]

**Review by emiliodcsa — APPROVED** [review]

File: jit/v2/JIT_v2.0.0.yaml [hunk 1]
```diff
@@ -41,16 +41,26 @@ info:
...
 
-    To request a **Port Call Service** the endPoint to use is:
+    ## How to create a Port Call Service
+    To request a **Port Call Service** do the following:
+     - Initiate a **Port Call** by calling the
 
-        PUT /v2/port-call-services
+            PUT /v2/port-calls/{portCallID}
+
+     - Initiate a **Terminal Call** and link it to the **Port Call** create above by calling
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:50]: shouldn't this be 'created'? same for the next one too.

**Review by emiliodcsa — APPROVED** [review]

**Review by emiliodcsa — APPROVED** [review]

File: jit/v2/JIT_v2.0.0.yaml [hunk 2]
```diff
@@ -146,18 +154,17 @@ paths:
...
                     **NB**: `errorCode` not yet standardized by DCSA. Value `7003` is just a "random example"
                   value:
                     httpMethod: PUT
-                    requestUri: /v2/port-call-services/085a3207-5e45-49cf-8e1b-f8442beaf545
+                    requestUri: /v2/port-calls/085a3207-5e45-49cf-8e1b-f8442beaf545
                     statusCode: 500
                     statusCodeText: Internal Server Error
-                    statusCodeMessage: |
-                      Internal Server Error occurred while processing Port Call Service
+                    statusCodeMessage: Internal Server Error occurred while processing Port Call Service
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:167]: Port call service -> Port call

File: jit/v2/JIT_v2.0.0.yaml [hunk 3]
```diff
@@ -174,30 +181,139 @@ paths:
...
+                    providerCorrelationReference: 4426d965-0dd8-4005-8c63-dc68b01c4962
+                    errorDateTime: '2024-11-21T09:41:00Z'
+                    errors:
+                      - errorCode: 7004
+                        property: isFYI
+                        errorCodeText: wrong property type
+                        errorCodeMessage: isFYI should be a boolean. '12' provided as value
+        '404':
+          description: |
+            In case the provider does not know of the `portCallID` used in the request (this could be because of a `PUT` request that has not finished processing or simply because the resource does not exist) - it is possible for the provider to reject the requests by returning a `404` (Not Found).
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:243]: Provider or Consumer?

File: jit/v2/JIT_v2.0.0.yaml [hunk 4]
```diff
@@ -174,30 +181,139 @@ paths:
...
+                    providerCorrelationReference: 4426d965-0dd8-4005-8c63-dc68b01c4962
+                    errorDateTime: '2024-11-21T09:41:00Z'
+                    errors:
+                      - errorCode: 7004
+                        property: isFYI
+                        errorCodeText: wrong property type
+                        errorCodeMessage: isFYI should be a boolean. '12' provided as value
+        '404':
+          description: |
+            In case the provider does not know of the `portCallID` used in the request (this could be because of a `PUT` request that has not finished processing or simply because the resource does not exist) - it is possible for the consumer to reject the requests by returning a `404` (Not Found).
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:243]: In case the consumer does not know?

File: jit/v2/JIT_v2.0.0.yaml [hunk 5]
```diff
@@ -174,30 +181,139 @@ paths:
...
+                    requestUri: /v2/port-calls/085a3207-5e45-49cf-8e1b-f8442beaf545/omit
+                    statusCode: 404
+                    statusCodeText: Not Found
+                    statusCodeMessage: portCallID not found
+                    providerCorrelationReference: 4426d965-0dd8-4005-8c63-dc68b01c4962
+                    errorDateTime: '2024-09-04T09:41:00Z'
+                    errors:
+                      - errorCode: 7006
+                        errorCodeText: portCallID Not Found
+                        errorCodeMessage: The Port Call Service does not exist
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:270]: is it Port call service or Port call?

File: jit/v2/JIT_v2.0.0.yaml [hunk 6]
```diff
@@ -174,30 +181,139 @@ paths:
...
+        '500':
+          description: |
+            In case a server error occurs in provider system a `500` (Internal Server Error) is returned
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:273]: Consumer

File: jit/v2/JIT_v2.0.0.yaml [hunk 7]
```diff
@@ -174,30 +181,139 @@ paths:
...
+  '/v2/port-calls/{portCallID}/omit':
+    post:
       parameters:
-        - $ref: '#/components/parameters/portCallIDQueryParam'
-        - $ref: '#/components/parameters/terminalCallIDQueryParam'
-        - $ref: '#/components/parameters/portCallServiceIDQueryParam'
+        - $ref: '#/components/parameters/portCallIDPathParam'
+        - $ref: '#/components/parameters/Api-Version-Major'
+      tags:
+        - Port Call Service consumer
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:190]: 'C'onsumer? -- just that it looks aligned with Port Call Service

File: jit/v2/JIT_v2.0.0.yaml [hunk 8]
```diff
@@ -301,58 +417,24 @@ paths:
...
-                    errorDateTime: '2024-09-04T09:41:00Z'
-                    errors:
-                      - errorCode: 7006
-                        errorCodeText: portCallServiceID Not Found
-                        errorCodeMessage: |
-                          The Port Call Service does not exist
+                        errorCodeMessage: carrierServiceCode must be provided as part of a Terminal Call
         '500':
           description: |
             In case a server error occurs in provider system a `500` (Internal Server Error) is returned
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:440]: provider -> consumer

File: jit/v2/JIT_v2.0.0.yaml [hunk 9]
```diff
@@ -366,22 +448,21 @@ paths:
               examples:
                 internalServerErrorExample:
                   summary: |
-                    Internal Server Error while processing Port Call Service
+                    Internal Server Error while processing Terminal Call
                   description: |
                     An Internal Server Error has occurred - the consumer should contact {provider-support} and provide the `providerCorrelationReference` (in the example this is `4426d965-0dd8-4005-8c63-dc68b01c4962`)
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:453]: consumer and provider

File: jit/v2/JIT_v2.0.0.yaml [hunk 10]
```diff
@@ -366,22 +448,21 @@ paths:
...
+                    **NB**: `errorCode` not yet standardized by DCSA. Value `7003` is just a "random example"
                   value:
                     httpMethod: PUT
-                    requestUri: /v2/port-call-services/085a3207-5e45-49cf-8e1b-f8442beaf545/estimated-timestamp
+                    requestUri: /v2/terminal-calls/085a3207-5e45-49cf-8e1b-f8442beaf545
                     statusCode: 500
                     statusCodeText: Internal Server Error
-                    statusCodeMessage: |
-                      Internal Server Error occurred while processing Port Call Service
+                    statusCodeMessage: Internal Server Error occurred while processing Port Call Service
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:461]: Terminal call?

File: jit/v2/JIT_v2.0.0.yaml [hunk 11]
```diff
@@ -432,29 +514,27 @@ paths:
...
                       - errorCode: 7004
                         property: isFYI
                         errorCodeText: wrong property type
-                        errorCodeMessage: |
-                          isFYI should be a boolean. '12' provided as value
+                        errorCodeMessage: isFYI should be a boolean. '12' provided as value
         '404':
           description: |
-            In case the provider does not know of the `portCallServiceID` used in the request (this could be because of a `PUT` request that has not finished processing or simply because the resource does not exist) - it is possible for the provider to reject the requests by returning a `404` (Not Found).
+            In case the provider does not know of the `terminalCallID` used in the request (this could be because of a `PUT` request that has not finished processing or simply because the resource does not exist) - it is possible for the consumer to reject the requests by returning a `404` (Not Found).
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:537]: consumer in the first sentence

File: jit/v2/JIT_v2.0.0.yaml [hunk 12]
```diff
@@ -467,23 +547,21 @@ paths:
                   summary: |
                     Not Found request
                   description: |
-                    The provided `portCallServiceID` cannot be found. This can be because a `PUT` request has not been finished processing or because the `portCallServiceID` does not exist in the provider system.
+                    The provided `terminalCallID` cannot be found. This can be because a `PUT` request has not been finished processing or because the `terminalCallID` does not exist in the provider system.
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:550]: consumer system?

File: jit/v2/JIT_v2.0.0.yaml [hunk 13]
```diff
@@ -826,29 +1004,27 @@ paths:
...
-                        errorCodeMessage: |
-                          isFYI should be a boolean. '12' provided as value
+                      - errorCode: 7003
+                        property: classifierCode
+                        errorCodeText: mandatory property missing
+                        errorCodeMessage: classifierCode must be provided as part of a Timestamp
         '404':
           description: |
-            In case the provider does not know of the `portCallServiceID` used in the request (this could be because of a `PUT` request that has not finished processing or simply because the resource does not exist) - it is possible for the provider to reject the requests by returning a `404` (Not Found).
+            In case the provider does not know of the `portCallServiceID` used in the request, it is possible for the provider to reject the requests by returning a `404` (Not Found). This could be because of a `PUT` request, for creating/initiating the Port Call Service, that has not finished processing or simply because the resource does not exist.
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:1027]: timestampID

File: jit/v2/JIT_v2.0.0.yaml [hunk 14]
```diff
@@ -897,33 +1071,419 @@ paths:
...
+              examples:
+                internalServerErrorExample:
+                  summary: |
+                    Internal Server Error while processing Port Call Service
+                  description: |
+                    An Internal Server Error has occurred - the consumer should contact {provider-support} and provide the `providerCorrelationReference` (in the example this is `4426d965-0dd8-4005-8c63-dc68b01c4962`)
+
+                    **NB**: `errorCode` not yet standardized by DCSA. Value `7005` is just a "random example"
+                  value:
+                    httpMethod: PUT
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:1144]: GET and request uri too.

File: jit/v2/JIT_v2.0.0.yaml (hunk 4 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml]: changed to `implementor`

File: jit/v2/JIT_v2.0.0.yaml (hunk 5 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml:270]: Correct - changed

File: jit/v2/JIT_v2.0.0.yaml (hunk 6 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml]: Changed to `caller` and `implementor`

File: jit/v2/JIT_v2.0.0.yaml (hunk 7 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml]: Modified it here plus all other places to: **Port Call Service - Consumer** (same for Provider tags)

File: jit/v2/JIT_v2.0.0.yaml (hunk 8 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml]: changed to caller and implementor

File: jit/v2/JIT_v2.0.0.yaml [hunk 15]
```diff
@@ -893,60 +1067,446 @@ paths:
...
+              examples:
+                internalServerErrorExample:
+                  summary: |
+                    Internal Server Error while processing Port Call Service
+                  description: |
+                    An Internal Server Error has occurred - the caller should contact {implementor-support} and provide the `providerCorrelationReference` (in the example this is `4426d965-0dd8-4005-8c63-dc68b01c4962`)
+
+                    **NB**: `errorCode` not yet standardized by DCSA. Value `7005` is just a "random example"
+                  value:
+                    httpMethod: PUT
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:1335]: Get v2/vessel-statuses

File: jit/v2/JIT_v2.0.0.yaml [hunk 16]
```diff
@@ -893,60 +1067,446 @@ paths:
...
+    post:
+      parameters:
+        - $ref: '#/components/parameters/portCallServiceIDPathParam'
+        - $ref: '#/components/parameters/Api-Version-Major'
+      tags:
+        - Port Call Service provider
+      operationId: decline-port-call-service
+      summary: Decline a Port Call service
+      description: |
+        Consumer declining a **Port Call Service**.
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:1366]: Provider

File: jit/v2/JIT_v2.0.0.yaml (hunk 10 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml]: Fixed along with a lot of other places having the same error :-)

File: jit/v2/JIT_v2.0.0.yaml (hunk 13 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml]: This was on purpose as it is referring to the portCallServiceID inside the payload ;)

File: jit/v2/JIT_v2.0.0.yaml (hunk 14 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml]: fixed + other places as well

File: jit/v2/JIT_v2.0.0.yaml (hunk 15 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml]: fixed

File: jit/v2/JIT_v2.0.0.yaml [hunk 17]
```diff
@@ -1163,21 +1725,41 @@ components:
...
+        example: 0342254a-5927-4856-b9c9-aa12e7c00563
+    
     # Pagination query parameters
 
     limitQueryParam:
       in: query
       name: limit
       description: |
-        Specifies the maximum number of `Port Call Service` objects to return.
+        Specifies the maximum number of objects in the resultset.
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:1762]: result set

File: jit/v2/JIT_v2.0.0.yaml [hunk 18]
```diff
@@ -1382,6 +1921,28 @@ components:
...
+          description: |
+            The carrier specific code of the service for which the schedule details are published.
+          example: FE1
+        universalServiceReference:
+          type: string
+          pattern: ^SR\d{5}[A-Z]$
+          minLength: 8
+          maxLength: 8
+          description: |
+            A global unique (Vessel-)service reference, as per DCSA standard, agreed by VSA partners for the (Vessel-)service. The (Vessel-)service reference must match the regular expression pattern: `SR\d{5}[A-Z]`. The letters `SR` followed by `5 digits`, followed by a checksum-character as a capital letter from `A to Z`.
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:1944]: As per Emilio the USR description should be - "A global unique service reference, as per DCSA standard, agreed by VSA partners for the service. The service reference must match the regular expression pattern: SR\d{5}[A-Z]. The letters SR followed by 5 digits, followed by a checksum-character as a capital letter from A to Z."

File: jit/v2/JIT_v2.0.0.yaml (hunk 16 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml:1366]: No this is correct as the decline is the opposite way (compared to cancel)

File: jit/v2/JIT_v2.0.0.yaml (hunk 17 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml:1762]: fixed

File: jit/v2/JIT_v2.0.0.yaml (hunk 18 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml:1944]: @emiliodcsa I guess we still need to distinguish between Services that are for Vessels (which is the case here) and Services in the context of JIT-Services...? Should I change the wording?

File: jit/v2/JIT_v2.0.0.yaml [hunk 19]
```diff
@@ -1701,54 +2159,341 @@ components:
...
+            - `LOADING_CARGO_OPERATIONS` (Loading cargo operations)
+            - `LASHING` (Lashing)
+            - `SAFETY` (Safety)
+            - `ANCHORAGE_OPERATIONS` (Anchorage Operations)
+            - `SHORE_POWER` (ShorePower)
+
+            **Condition:** `ANCH`, when used, should be used with the following `portCallServiceType`:
+            - `ANCHORAGE` (Anchorage)
+            - `ANCHORAGE_OPERATIONS` (Anchorage Operations)
+          enum:
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:2196]: Are these fixed values, should this be an enum or pseudo enum?

File: jit/v2/JIT_v2.0.0.yaml (hunk 18 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml:1944]: @emiliodcsa says only these three values in this version (any updates will require a new major release...)

File: jit/v2/JIT_v2.0.0.yaml (hunk 19 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml:2196]: @emiliodcsa says only these three values in this version (any updates will require a new major release...)

File: jit/v2/JIT_v2.0.0.yaml (hunk 18 above)
Comment (HenrikHL) [reply jit/v2/JIT_v2.0.0.yaml:1944]: We will keep as it is to make sure it is a vessel service...

File: jit/v2/JIT_v2.0.0.yaml [hunk 20]
```diff
@@ -65,44 +75,44 @@ info:
...
           description: |
-            A new Port Call Service accepted
+            A new or updated Port Call accepted
           headers:
             API-Version:
               $ref: '#/components/headers/API-Version'
         '400':
           description: |
-            In case creating a new `Port Call Service` fails schema validation a `400` (Bad Request) is returned
+            In case creating a new `Port Call` fails schema validation a `400` (Bad Request) is returned
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:115]: can we add a ',' after schema validation? In some places it is added already.

File: jit/v2/JIT_v2.0.0.yaml [hunk 21]
```diff
@@ -41,16 +41,26 @@ info:
...
 
-    To request a **Port Call Service** the endPoint to use is:
+    ## How to create a Port Call Service
+    To request a **Port Call Service** do the following:
+     - Create a **Port Call** by calling the
 
-        PUT /v2/port-call-services
+            PUT /v2/port-calls/{portCallID}
+
+     - Create a **Terminal Call** and link it to the **Port Call** create above by calling
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:50]: **Port Call** created

File: jit/v2/JIT_v2.0.0.yaml [hunk 22]
```diff
@@ -208,24 +324,24 @@ paths:
...
               $ref: '#/components/headers/API-Version'
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/PortCallServices'
+                $ref: '#/components/schemas/PortCalls'
         '500':
           description: |
-            In case a server error occurs in provider system a `500` (Internal Server Error) is returned
+            In case a server error occurs in backend system a `500` (Internal Server Error) is returned
```
Comment (palatsangeetha) [inline jit/v2/JIT_v2.0.0.yaml:344]: Can we change this to implementor, rest everywhere it is implementor

**Review by palatsangeetha — APPROVED** [review]
The changes looks good to me.

**Pull request merged**